File: part2a/compute-probas.c
/**********************************************************
 * calcul_probas.c : program to compute the probabilities *
 * of the S-Boxes, needed to assess the quality of the    *
 * differential paths                                     *
 **********************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int s[16] = {0xe,0x4,0xd,0x1,0x2,0xf,0xb,0x8,
	     0x3,0xa,0x6,0xc,0x5,0x9,0x0,0x7};


int probas[16][16];


void compute_probas () {
  int x, deltaIN, deltaOUT;

  for(deltaIN=0; deltaIN<=15; deltaIN++)
  {
    for(deltaOUT=0; deltaOUT<=15; deltaOUT++)
    {
      probas[deltaIN][deltaOUT]=0;
    }
  }

  for(deltaIN=0; deltaIN<=15; deltaIN++)
  {
    for(x=0; x<=15; x++)
    {
      deltaOUT=s[x] ^ s[x ^ deltaIN];
      probas[deltaIN][deltaOUT]++;
    }
  }
}


void print_probas () {
  int i, delta_in,delta_out;
  char ligne[16*3 + 4 + 1];

  // Print the first line
  printf ("   |");
  for (delta_out=0; delta_out<16; delta_out++)
    printf (" %x ", delta_out);
  printf ("\n");

  for (i=0; i<16*3 + 4; i++)
    ligne[i]='-';
  ligne[16*3 + 4] = 0;
  printf ("%s\n", ligne);
  
  for (delta_in=0; delta_in<16; delta_in++) {
    printf (" %x |", delta_in);
    for (delta_out=0; delta_out<16; delta_out++)
      printf ("%2d ", probas[delta_in ][delta_out]);
    printf ("\n");
  }
}


int main () {
  compute_probas ();
  print_probas ();
  return 0;
}

File: part2a/compute-probas.py
#!/usr/bin/env python3

###########################################################
# calcul_probas.py : program to compute the probabilities #
# of the S-Boxes, needed to assess the quality of the     #
# differential paths                                      #
###########################################################

import sys


s = [0xE, 0x4, 0xD, 0x1, 0x2, 0xF, 0xB, 0x8, 0x3, 0xA, 0x6, 0xC, 0x5, 0x9, 0x0, 0x7]

probas = []


def compute_probas():
    global probas, s



def print_probas():
    # Print the first line
    sys.stdout.write("   |")
    for delta_out in range(16):
        sys.stdout.write(" %x " % delta_out)
    sys.stdout.write("\n")

    sys.stdout.write("-" * (16 * 3 + 4))
    sys.stdout.write("\n")

    for delta_in in range(16):
        sys.stdout.write(" %x |" % delta_in)
        for delta_out in range(16):
            sys.stdout.write("%2d " % probas[delta_in][delta_out])
        sys.stdout.write("\n")


compute_probas()
print_probas()
